osutil/sync_dir.go

```diff
@@ -0,0 +1,204 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"syscall"
+)
+
+// Janitor looks after a specific directory, making sure it has exactly what is expected.
+//
+// Technically the janitor looks at a subset of files in a given directory. On
+// demand, the janitor scans the directory and looking at contents and
+// permissions of files matching a given pattern. At the same time the janitor
+// has an oracle/helper that tells it exactly what should be in each of the
+// files, which files should exist, etc. The janitor uses this information to
+// correct any discrepancies.
+//
+// After checking everything the janitor notifies whoever controls it about the
+// changes made. Those changes can be used to follow up with some additional
+// tasks that are beyond the scope of the janitor, e.g. to run some programs.
+//
+// Note that Janitor doesn't use any filesystem observation APIs. All changes
+// are monitored on demand.
+type Janitor struct {
```

Comment:
I just noticed that Gustavo commented earlier and suggested to move it here. Its fine (I think) however I would prefer more specific names/functions (`EnforceDirectory()` or `SyncDirectory` (but its more like a force instead of a sync but I won't argue over this, just my personal feeling).